schema/representations.md

```diff
@@ -0,0 +1,257 @@
+type Fizzlebop struct {
+	a String
+	b String
+} representation stringjoin {
+	join ":"
+}
+```
+
+### map stringpairs representation example
+
+Say we're doing something awfully like the mount options in an /etc/fstab file:
+
+```ipldsch
+## MountOptions serializes as a string of comma-separated "k=v" entries, e.g. "k1=v1,k2=v2".
+## This serial representation is limited (n.b. The domain of keys must exclude "=" and values must exclude ","!),
+## but we've chosen it because other parts of our application protocol are drastically simplified by limiting this to a string.
+type MountOptions map {String:String} representation stringpairs {
+	innerDelim "="
+	entryDelim ","
+}
+```
+
+### union keyed representation example
+
+```ipldsch
+type MyKeyedUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation keyed
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKeyedUnion` (shown as JSON) is:
+
+```json
+{"foo": {"froz": true}}
+```
+
+This data would also match, as the other type:
+
+```json
+{"bar": 12}
+```
+
+### union kinded representation example
+
+```ipldsch
+type MyKindedUnion union {
+	| Foo map
+	| Bar int
+} representation kinded
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyKindedUnion` (shown as JSON) is:
+
+```json
+{"froz": true}
+```
+
+This data would also match, as the other type:
+
+```json
+12
+```
+
+Note how kinded unions don't introduce any kind of wrapping map in the
+serialized form at all;
+contrast this with each of the other union representation strategies, all
+of which use at least one layer of map in their representation.
+
+Note that the syntax used in the type declaration is different for kinded
+unions versus other union representations!
+Kinded unions, unlike all other representation strategies, don't have
+any kind of string keywords at all.  Instead, they encode all of their
+discriminant information in the kind of the data itself!
+Correspondingly, the tagging information in the type definition uses kind
+keywords (unquoted) instead of quoted strings.
+
+Note that the kinds valid in a kinded union are all Data Model-layer
+*representation kinds*.  E.g. `string` and `map` are acceptable here;
+`struct` is not.  The kind that a union member type is tagged with must
+match that type's representation kind or the schema is invalid.
+
+### union envelope representation example
+
+```ipldsch
+type MyEnvelopeUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation kinded {
+	discriminantKey "tag"
+	contentKey "msg"
+}
+
+type Foo struct { froz Bool }
+type Bar int
+```
+
+Some data matching `MyEnvelopeUnion` (shown as JSON) is:
+
+```json
+{"tag":"foo", "msg":{"froz":true}}
+```
+
+This data would also match, as the other type:
+
+```json
+{"tag":"bar", "msg":12}
+```
+
+### union inline representation example
+
+```ipldsch
+type MyInlineUnion union {
+	| Foo "foo"
+	| Bar "bar"
+} representation kinded {
```

Comment:
`} representation inline {`

the schema-schema skips `discriminantKey ` and uses a quote string instead of this block, so it would be `} representation inline "tag"`. Is that intended to be supported as a shortcut as well as this long-form block?

Reply:
> intended to be supported as a shortcut

Unsure.  I think that line in the schema-schema might be an artifact of an earlier round of drafting.  The shortcut might be nice (if, indeed, we're sure it'll often be a single-arg use); otoh, the parser would be more consistent if we didn't add such shortcuts.  I think we're still open to going either way, and as someone who's written a grammar, I think your opinion counts more than mine here.